File: src/tracehub/streaming.py
"""
Script: streaming.py
Created: 2026-02-22
Purpose: SSE subscriber management for real-time trace streaming
Keywords: streaming, sse, subscribers, realtime, tracehub
Status: active
Prerequisites:
  - asyncio
Changelog:
  - 2026-02-22: Extracted from server.py during package refactor
See-Also: .claude/CLAUDE.md ##Управление скриптами
"""

import asyncio
import time

from .config import _subscribers, _subscriber_timestamps
from .models import TraceEntry


async def subscribe(correlation_id: str) -> asyncio.Queue:
    """Subscribe to real-time updates for a correlation ID."""
    if correlation_id not in _subscribers:
        _subscribers[correlation_id] = []
    queue: asyncio.Queue = asyncio.Queue(maxsize=100)
    _subscribers[correlation_id].append(queue)
    _subscriber_timestamps[correlation_id] = time.time()
    return queue


def unsubscribe(correlation_id: str, queue: asyncio.Queue):
    """Unsubscribe from updates."""
    if correlation_id in _subscribers:
        try:
            _subscribers[correlation_id].remove(queue)
            if not _subscribers[correlation_id]:
                del _subscribers[correlation_id]
        except ValueError:
            pass


async def notify_subscribers(trace: TraceEntry):
    """Notify all subscribers for a correlation ID."""
    corr_id = trace.correlation_id
    if corr_id in _subscribers:
        dead_queues = []
        for queue in _subscribers[corr_id]:
            try:
                queue.put_nowait(trace)
            except asyncio.QueueFull:
                dead_queues.append(queue)
            except Exception:
                dead_queues.append(queue)
        for q in dead_queues:
            try:
                _subscribers[corr_id].remove(q)
            except ValueError:
                pass
        if not _subscribers[corr_id]:
            del _subscribers[corr_id]
            _subscriber_timestamps.pop(corr_id, None)


async def cleanup_stale_subscribers():
    """Remove subscriber entries older than 5 minutes with no active queues."""
    cutoff = time.time() - 300
    stale = [
        cid for cid, ts in _subscriber_timestamps.items()
        if ts < cutoff
    ]
    for cid in stale:
        if cid in _subscribers:
            del _subscribers[cid]
        _subscriber_timestamps.pop(cid, None)

File: src/services/streaming.ts
import type { TraceEntry } from "../lib/types";

// =============================================================================
// SSE subscriber management — port of Python streaming.py
// =============================================================================

interface Subscriber {
	writer: WritableStreamDefaultWriter;
	addedAt: number;
}

/** corrId -> set of subscriber writers */
const _subscribers = new Map<string, Set<Subscriber>>();

/** corrId -> last activity timestamp (for stale cleanup) */
const _subscriberTimestamps = new Map<string, number>();

/**
 * Subscribe to real-time trace updates for a correlation ID.
 * Returns a Subscriber handle for cleanup.
 */
export function subscribe(
	corrId: string,
	writer: WritableStreamDefaultWriter,
): Subscriber {
	if (!_subscribers.has(corrId)) {
		_subscribers.set(corrId, new Set());
	}
	const sub: Subscriber = { writer, addedAt: Date.now() };
	_subscribers.get(corrId)!.add(sub);
	_subscriberTimestamps.set(corrId, Date.now());
	return sub;
}

/**
 * Unsubscribe a specific writer from a correlation ID.
 */
export function unsubscribe(corrId: string, sub: Subscriber): void {
	const subs = _subscribers.get(corrId);
	if (!subs) return;
	subs.delete(sub);
	if (subs.size === 0) {
		_subscribers.delete(corrId);
		_subscriberTimestamps.delete(corrId);
	}
}

/**
 * Notify all subscribers for a trace's correlation ID.
 * Dead writers are removed automatically.
 */
export function notifySubscribers(trace: TraceEntry): void {
	const corrId = trace.correlation_id;
	const subs = _subscribers.get(corrId);
	if (!subs || subs.size === 0) return;

	const data = `data: ${JSON.stringify(trace)}\n\n`;
	const encoder = new TextEncoder();
	const encoded = encoder.encode(data);
	const deadSubs: Subscriber[] = [];

	for (const sub of subs) {
		try {
			sub.writer.write(encoded).catch(() => {
				deadSubs.push(sub);
			});
		} catch {
			deadSubs.push(sub);
		}
	}

	for (const dead of deadSubs) {
		subs.delete(dead);
	}

	if (subs.size === 0) {
		_subscribers.delete(corrId);
		_subscriberTimestamps.delete(corrId);
	}

	_subscriberTimestamps.set(corrId, Date.now());
}

/**
 * Remove subscriber entries older than 5 minutes with no recent activity.
 */
export function cleanupStaleSubscribers(): void {
	const cutoff = Date.now() - 300_000; // 5 min
	const stale: string[] = [];

	for (const [corrId, ts] of _subscriberTimestamps) {
		if (ts < cutoff) {
			stale.push(corrId);
		}
	}

	for (const corrId of stale) {
		const subs = _subscribers.get(corrId);
		if (subs) {
			for (const sub of subs) {
				try {
					sub.writer.close();
				} catch {
					// already closed
				}
			}
		}
		_subscribers.delete(corrId);
		_subscriberTimestamps.delete(corrId);
	}
}

/**
 * Return subscriber statistics for /stats endpoint.
 */
export function getSubscriberStats(): {
	activeCorrelations: number;
	totalQueues: number;
} {
	let totalQueues = 0;
	for (const subs of _subscribers.values()) {
		totalQueues += subs.size;
	}
	return {
		activeCorrelations: _subscribers.size,
		totalQueues,
	};
}
